test/coretasks/test_coretasks_sasl.py

```diff
@@ -344,3 +353,201 @@ def test_sasl_nak(botfactory: BotFactory, tmpconfig) -> None:
         'CAP END',
         'QUIT :Error negotiating capabilities.',
     )
+
+
+def test_sasl_bad_method(mockbot: Sopel, caplog: pytest.LogCaptureFixture):
+    """Verify the bot behaves when configured with an unsupported SASL method."""
+    mockbot.settings.core.auth_method = "sasl"
+    mockbot.settings.core.auth_target = "SCRAM-MD4"
+    mockbot.on_message("CAP * LS :sasl")
+    mockbot.on_message("CAP TestBot ACK :sasl")
+    assert mockbot.backend.message_sent == rawlist(
+        "CAP REQ :sasl",
+        "CAP END",
+    )
+    with caplog.at_level(ERROR):
+        mockbot.on_message("AUTHENTICATE +")
+    assert '"SCRAM-MD4" is not supported' in caplog.text
+
+
+def test_sasl_plain_auth(mockbot: Sopel):
+    """Verify the bot performs SASL PLAIN auth correctly."""
+    mockbot.settings.core.auth_method = "sasl"
+    mockbot.settings.core.auth_target = "PLAIN"
+    mockbot.on_message("CAP * LS :sasl")
+    mockbot.on_message("CAP TestBot ACK :sasl")
+    assert mockbot.backend.message_sent == rawlist(
+        "CAP REQ :sasl",
+        "AUTHENTICATE PLAIN",
+    )
+    mockbot.on_message("AUTHENTICATE +")
+    assert (
+        len(mockbot.backend.message_sent) == 3
+        and mockbot.backend.message_sent[-1]
+        == rawlist("AUTHENTICATE VGVzdEJvdABUZXN0Qm90AHNlY3JldA==")[0]
+    )
+    mockbot.on_message(
+        "900 TestBot test!test@test TestBot :You are now logged in as TestBot"
+    )
+    mockbot.on_message("903 TestBot :SASL authentication succeeded")
+    assert (
+        len(mockbot.backend.message_sent) == 4
+        and mockbot.backend.message_sent[-1] == rawlist("CAP END")[0]
+    )
+
+
+def test_sasl_scram_sha_256_auth(mockbot: Sopel):
+    """Verify the bot performs SASL SCRAM-SHA-256 auth correctly."""
+    mech = ScramMechanism()
+    salt, stored_key, server_key, iter_count = mech.make_auth_info(
+        "secret", iteration_count=5000
+    )
+    scram_server = mech.make_server(
+        lambda x: (salt, stored_key, server_key, iter_count)
+    )
+
+    mockbot.settings.core.auth_method = "sasl"
+    mockbot.settings.core.auth_target = "SCRAM-SHA-256"
+    mockbot.on_message("CAP * LS :sasl")
+    mockbot.on_message("CAP TestBot ACK :sasl")
+    assert mockbot.backend.message_sent == rawlist(
+        "CAP REQ :sasl",
+        "AUTHENTICATE SCRAM-SHA-256",
+    )
+    mockbot.on_message("AUTHENTICATE +")
+
+    scram_server.set_client_first(
+        b64decode(mockbot.backend.message_sent[-1].split(b" ")[-1]).decode("utf-8")
+    )
```

Comment:
The `[-1]` here is an assumption that I don't like. I'd rather see that there is, indeed, only the expected number of messages. And if possible, to check the content of said messages (it might not be possible, I don't know if SCRAM has a random part that we can't control in any way here).